src/orion/benchmark/task/profet/profet_task.py

```diff
@@ -0,0 +1,817 @@
+
+        self.device: torch.device
+        if isinstance(device, torch.device):
+            self.device = device
+        else:
+            self.device = torch.device(
+                device or ("cuda" if torch.cuda.is_available() else "cpu")
+            )
+
+        # TODO: Save those in the configuration dict?
+        self._np_rng: Optional[np.random.RandomState] = None
+        self._torch_rng_state: Optional[Tensor] = None
+        with self.seed_randomness():
+            if os.path.exists(self.checkpoint_file):
+                logger.info(
+                    f"Model has already been trained: loading it from file {self.checkpoint_file}."
+                )
+                self.net, self.h = load_task_network(self.checkpoint_file)
+            else:
+                logger.info(
+                    f"Checkpoint file {self.checkpoint_file} doesn't exist: re-training the model."
+                )
+                logger.info(f"Task hash params: {task_hash_params}")
+
+                self.checkpoint_file.parent.mkdir(exist_ok=True, parents=True)
+                # Need to re-train the meta-model and sample this task.
+                self.net, self.h = get_task_network(
+                    self.input_dir,
+                    self.benchmark,
+                    seed=self.seed,
+                    task_id=self.task_id,
+                    config=self.train_config,
+                )
+                save_task_network(
+                    self.checkpoint_file, self.benchmark, self.net, self.h
+                )
+
+        self.net = self.net.to(device=self.device)
+        self.net.eval()
+        self.h_tensor = torch.as_tensor(self.h, dtype=torch.float32, device=self.device)
+
+        self._space: Space = SpaceBuilder().build(self.get_search_space())
+        # IDEA: Could 'hack' the `sample` function of this Space instance so it always uses our
+        # numpy rng state.
+        # _original_sample = self._space.sample
+
+        # def seeded_sample(_, n_samples: int = 1, seed=None):
+        #     nonlocal self
+        #     if seed is None:
+        #         seed = self._np_rng
+        #     return _original_sample(n_samples=n_samples, seed=seed)
+
+        # self._space.sample = seeded_sample
+
+        self.name = f"profet.{type(self).__qualname__.lower()}_{self.task_id}"
+
+    @contextmanager
+    def seed_randomness(self):
+        """Context manager used to seed the portions of the code that use randomness, without
+        affecting the global state.
+        """
+        devices = [] if self.device.type == "cpu" else [self.device]
+        if self._np_rng is None:
+            # Initialize the numpy random state.
+            self._np_rng = RandomState(seed=self.seed)
+
+        start_np_state = np.random.get_state()
+        # Temporarily set the global random state to that of our RNG.
+        np.random.set_state(self._np_rng.get_state())
```

Comment:
Looking at the code again, I don't understand why the global numpy RNG is seeded temporarily here. It's not used in the with-clause right?

Reply:
Reworked the seeding quite a bit, Model training / loading / forward pass is reproducible.

Reply:
And as mentioned in [the other comment](https://github.com/Epistimio/orion/pull/632#discussion_r781410294), both training and loading have an effect on the global torch RNG, since when loading, it first creates a new Model, and then loads the state_dict from the checkpoint. Creating a new model affects the global pytorch RNG when initializing the weights, so both need to be seeded here.

Reply:
Alright, thanks for looking into all this!